api/nodes/views.py

```diff
@@ -1367,6 +1371,168 @@ def get_object(self):
+        ReadOnlyIfRegistration,
+        base_permissions.TokenHasScope,
+    )
+
+    serializer_class = NodeAddonSettingsSerializer
+    view_category = 'nodes'
+    view_name = 'node-addon-detail'
+
+    def get_object(self):
+        addon = self.kwargs['provider']
+        if addon in ADDONS_OAUTH:
+            addon_settings = ADDONS_AVAILABLE_DICT[addon].settings_models.get('node', None)
+            obj = None
+            enabled = False
+            try:
+                obj = addon_settings.find(Q('owner', 'eq', self.get_node()) & Q('deleted', 'eq', False))[0]
+                enabled = True
+            except IndexError:
+                pass
+            return {
+                '_id': addon,
+                'enabled': enabled,
+                'settings': obj
+            }
+        raise NotFound('Requested addon unavailable.')
+
+class NodeAddonFolderList(JSONAPIBaseView, generics.ListAPIView, NodeMixin):
+    """
+    TODO: docs
+    """
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        ContributorOrPublic,
+        ExcludeRetractions,
+        base_permissions.TokenHasScope,
+    )
+
+    serializer_class = NodeAddonFolderSerializer
+    view_category = 'nodes'
+    view_name = 'node-addon-folders'
+
+    def _to_hgrid(self, item, addon, path):
+        """
+        :param item: contents returned from Google Drive API
+        :return: results formatted as required for Hgrid display
+        """
+        # quote fails on unicode objects with unicode characters
+        # covert to str with .encode('utf-8')
+        import os
+        from urllib import quote
+        # TODO: MVP deadlines make code bad
+        safe_name = quote(item['title'].encode('utf-8'), safe='')
+        path = os.path.join(path, safe_name)
+
+        serialized = {
+            'path': path,
+            'folder_id': item['id'],
+            'kind': 'folder',
+            'name': safe_name,
+            'provider': addon
+        }
+        return serialized
+
+    def get_queryset(self):
+        # TODO: refactor this/NS models to be generalizable
+        addon = self.kwargs['provider']
+        if addon != 'googledrive':
+            raise MethodNotAllowed('Requested addon unavailable.')
+
+        node = self.get_node()
+        node_addon = node.get_addon(addon)
+
+        if not node_addon:
+            raise NotFound('Requested addon not enabled.')
+
+        path = self.request.query_params.get('path', '')
+        folder_id = self.request.query_params.get('folder_id', 'root')
+
+        # TODO: MVP deadlines make code bad
+        from website.addons.base.exceptions import InvalidAuthError
+        from website.addons.googledrive.client import GoogleDriveClient
+
+        try:
+            access_token = node_addon.fetch_access_token()
+        except InvalidAuthError:
+            raise PermissionDenied('Invalid credentials.')
+
+        client = GoogleDriveClient(access_token)
```

Comment:
Discussed in person. For this iteration GoogleDrive specific code is acceptable.